File: trading_agent/risk.py
from __future__ import annotations

from dataclasses import dataclass

from trading_agent.config import RiskConfig
from trading_agent.models import Order, Side


@dataclass(frozen=True)
class RiskDecision:
    approved: bool
    reason: str


class RiskManager:
    def __init__(self, config: RiskConfig) -> None:
        self.config = config
        self.realized_pnl_today = 0.0

    def approve(self, order: Order, cash: float, positions: dict[str, int]) -> RiskDecision:
        order_value = order.quantity * order.price
        current_position = positions.get(order.symbol, 0)

        if order.quantity <= 0:
            return RiskDecision(False, "quantity must be positive")
        if order_value > self.config.max_order_value:
            return RiskDecision(False, "order value exceeds max_order_value")
        if self.realized_pnl_today <= -abs(self.config.max_daily_loss):
            return RiskDecision(False, "max_daily_loss reached")

        if order.side == Side.BUY:
            if order_value > cash:
                return RiskDecision(False, "not enough cash")
            new_position_value = (current_position + order.quantity) * order.price
            if new_position_value > self.config.max_position_value:
                return RiskDecision(False, "position value exceeds max_position_value")

        if order.side == Side.SELL:
            if not self.config.allow_short_selling and order.quantity > current_position:
                return RiskDecision(False, "short selling is disabled")

        return RiskDecision(True, "approved")

File: src/risk.js
import fs from "node:fs";
import path from "node:path";
import { Side } from "./models.js";

export class RiskManager {
  constructor(config) {
    this.config = config;
    this.ensureStopFile();
  }

  approve(order, account) {
    const orderValue = order.quantity * order.price;
    const currentPosition = account.positions.get(order.symbol) ?? 0;

    if (this.isEmergencyStopped()) return reject("emergency stop is active");
    if (order.quantity <= 0) return reject("quantity must be positive");
    if (orderValue > this.config.maxOrderValue) return reject("order value exceeds maxOrderValue");
    if (account.realizedPnlToday <= -Math.abs(this.config.maxDailyLoss)) return reject("maxDailyLoss reached");

    if (order.side === Side.BUY) {
      if (orderValue > account.cash) return reject("not enough cash");
      const newPositionValue = (currentPosition + order.quantity) * order.price;
      if (newPositionValue > this.config.maxPositionValue) return reject("position value exceeds maxPositionValue");
    }

    if (order.side === Side.SELL) {
      if (!this.config.allowShortSelling && order.quantity > currentPosition) {
        return reject("short selling is disabled");
      }
    }

    return { approved: true, reason: "approved" };
  }

  activateEmergencyStop(reason = "manual stop") {
    this.writeStopFile({ active: true, reason, updatedAt: new Date().toISOString() });
  }

  clearEmergencyStop() {
    this.writeStopFile({ active: false, reason: "resumed", updatedAt: new Date().toISOString() });
  }

  isEmergencyStopped() {
    try {
      const raw = JSON.parse(fs.readFileSync(this.config.emergencyStopFile, "utf8"));
      return Boolean(raw.active);
    } catch {
      return false;
    }
  }

  ensureStopFile() {
    if (fs.existsSync(this.config.emergencyStopFile)) return;
    this.writeStopFile({ active: false, reason: "initial", updatedAt: new Date().toISOString() });
  }

  writeStopFile(payload) {
    fs.mkdirSync(path.dirname(this.config.emergencyStopFile), { recursive: true });
    fs.writeFileSync(this.config.emergencyStopFile, JSON.stringify(payload, null, 2), "utf8");
  }
}

function reject(reason) {
  return { approved: false, reason };
}
